micropython/lora/lora/sx1276.py

```diff
@@ -0,0 +1,592 @@
+                _MODEM_CONFIG2_SYMB_TIMEOUT_MSB_MASK,
+                timeout_syms >> 8,
+            )
+            self._reg_write(_REG_SYMB_TIMEOUT_LSB, timeout_syms & 0xFF)
+
+        # Allocate the full FIFO for RX
+        self._reg_write(_REG_FIFO_ADDR_PTR, 0)
+        self._reg_write(_REG_FIFO_RX_BASE_ADDR, 0)
+        self._set_mode(
+            _OPMODE_MODE_RX_CONTINUOUS if timeout_ms is None else _OPMODE_MODE_RX_SINGLE
+        )
+
+        # will_irq if DIO0 and DIO1 both hooked up, or DIO0 and no timeout
+        return self._dio0 and (self._dio1 or timeout_ms is None)
+
+    def is_receive_completed(self):
+        # assumption here that _REG_IRQ_FLAGS_MASK is set correctly, so we don't have to check exactly which flags are set
+        return self._reg_read(_REG_IRQ_FLAGS)
+
+    def end_receive(self, timeout_ms):
+        self._irq_en = ()
+        if timeout_ms is None:
+            # We need to use continuous mode for a no-timeout receive, so stop receiving now
+            self._set_mode(_OPMODE_MODE_STDBY)
+
+        flags = self._reg_update(
+            _REG_IRQ_FLAGS,
+            _IRQ_RX_DONE | _IRQ_RX_TIMEOUT,
+            _IRQ_RX_DONE | _IRQ_RX_TIMEOUT,
+        )  # clear flags
+        if flags & _IRQ_RX_TIMEOUT:  # Modem timed out
+            return None
+
+        fifo = self._reg_read(_REG_RX_NB_BYTES)
+        message = bytearray(fifo)
+        self._reg_readinto(_REG_FIFO, message)
+
+        snr = struct.unpack("b", bytes([self._reg_read(_REG_PKT_SNR_VAL)]))[0]  # units: dB*4
+        rssi = self._reg_read(_REG_PKT_RSSI_VAL) - (
+            157 if self._reg_read(_REG_PA_CONFIG) & _PA_CONFIG_PASELECT_PA_BOOST_PIN else 164
+        )
+
+        return (message, snr, rssi)
+
+    def start_transmit(self, message):
+        self._set_mode(_OPMODE_MODE_STDBY)
+
+        if len(message) > _REG_MAX_PAYLOAD_LEN:
+            raise ValueError("message too long")
+
+        # Allocate the full FIFO for TX
+        self._reg_write(_REG_FIFO_ADDR_PTR, 0)
+        self._reg_write(_REG_FIFO_TX_BASE_ADDR, 0)
+
+        self._reg_write(_REG_PAYLOAD_LEN, len(message))
+
+        self._reg_write(_REG_FIFO, message)
+
+        self._reg_write(_REG_IRQ_FLAGS_MASK, 0xFF & ~(_IRQ_TX_DONE))
+
+        if self._dio0:
+            self._reg_update(
+                _REG_DIO_MAPPING1,
+                _DIO0_MAPPING_MASK << _DIO0_MAPPING_SHIFT,
+                1 << _DIO0_MAPPING_SHIFT,
+            )  # DIO0 = TXDone
+            self._irq_en = (self._dio0,)
+
+        self._set_mode(_OPMODE_MODE_TX)
+        return self._dio0 is not None  # will_irq if dio0 is set
+
+    # Use the same function, as setting _REG_IRQ_FLAGS_MASK ensures we only get relevanti nterrupts
+    is_transmit_completed = is_receive_completed
+
+    def end_transmit(self):
+        self._irq_en = ()
+        flags = self._reg_update(_REG_IRQ_FLAGS, _IRQ_TX_DONE, _IRQ_TX_DONE)  # clear flags
+        assert flags & _IRQ_TX_DONE
```

Comment:
Consistency check that we don't get here unless the `TX_DONE` interrupt flag was set.

Probably makes more sense in light of the other comments about begin_xxx/end_xxx/etc not being intended as public API functions. Sorry about mis-communicating this in the code!

Reply:
IMHO every assert should at the minimum have a comment that describes what an assertion failure indicates. Specifically here, I don't know whether it would signal invalid use of the end_transmit method by the caller or an unexpected situation WRT the state of the radio.